src/app/blog/frost-threshold-signatures/page.mdx

```diff
@@ -0,0 +1,346 @@
+-rw-r--r--  1 rklaehn  staff  134 Oct  1 18:51 25mzjgjlrcrma7wkm4l3fjv2afcs53cvmmyw3v2uwwt2dczsinaa.secret
+```
+
+This is all easy enough, but if the key shares are all on the same file system the scheme just adds complexity but no additional security.
+
+So we need a way to use the key shares from multiple machines that can be physically separated.
+
+Iroh is a library that can get you a fast and encrypted connection between any two devices anywhere in the world. So this should be easy, right? Let's see.
+
+### Remote operations - sign and cosign
+
+For using the key shares, there are two possible roles. The *signer* actively wants to sign a message, e.g. to publish it somewhere, but does not have the entire private key. Depending on the exact parameters for the key shares, it needs one more more co-signers.
+
+The *co-signer* is a little daemon that has one or more key share for which it can co-sign. For this exploration it will just wait for incoming co-sign requests and sign them.
+
+## The protocol
+
+The protocol looks like this
+
+<div className="not-prose">
+  <Image src="/blog/frost-threshold-signatures/protocol.jpeg" width={1222} height={1090} alt="protocol" />
+</div>
+
+- The signer sends a request to all configured co-signers to sign a message for a 32 byte public key.
+- Each co-signer that has a key share for the requested public key answers with a commitment and remembers a corresponding nonce.
+- The signer waits until it gets the required number of commitments. It then creates a signing package from all the commitments and the message and sends that to all co-signers that answered in the first round.
+- all co-signers sign the signing package and return a signature share.
+- as soon as the signer has enough signing shares, it can create a signature.
+- optionally we can validate that the signature is correct using the known ed public key.
+
+## Co-Signer
+
+The co-signer in this scheme acts as a server, so it needs to locally store its iroh keypair to have a stable node id. It also needs to publish discovery information. It does not, however, have to look up discovery information since it does not call other nodes.
+
+So this is how the endpoint setup looks like:
+
+```rust
+let discovery = PkarrPublisher::n0_dns(secret_key.clone());
+let endpoint = iroh_net::endpoint::Endpoint::builder()
+    .alpns(vec![COSIGN_ALPN.to_vec()])
+    .secret_key(secret_key)
+    .discovery(Box::new(discovery))
+    .bind()
+    .await?;
+```
+
+Once the endpoint is created, the co-signer needs to run a normal accept loop where it just handles incoming co-sign requests
+
+```rust
+while let Some(incoming) = endpoint.accept().await {
+    let data_path = data_path.clone();
+    tokio::spawn(async {
+        if let Err(cause) = handle_cosign_request(incoming, data_path).await {
+            tracing::error!("Error handling cosign request: {:?}", cause);
+        }
+    });
+}
+```
+
+Handling a request is described above.
```

Comment:
Oh right, and I remember being confused by this sentence.
Isn't it described *below*?

Reply:
I guess I must have moved stuff around.